### PR TITLE
9918 drop indices concurrently

### LINKED ISSUE #9537: Widgets/dataviews auto [invalid] index
I was reading [Problems with concurrent Postgres indexes (and how to solve them)](https://medium.com/carwow-product-engineering/problems-with-concurrent-postgres-indexes-and-how-to-solve-them-c57f7656c852) from [Postgres Weekly](http://postgresweekly.com/) and decided to see if I had invalid indexes in my installation.

I have some with the [`_auto_idx_` prefix](https://github.com/CartoDB/cartodb/blob/0e6d2c502931baffbdbf227ea773747030cd2045/app/services/carto/user_table_index_service.rb#L4). That's why I went with that title for the issue. But it could happen with anything using [Table.create_index with concurrent=true param](https://github.com/CartoDB/cartodb/blob/0e6d2c502931baffbdbf227ea773747030cd2045/app/models/table.rb#L1124).

I don't think this is critical but I wanted to report it just in case you think it's possible to improve it.

cc @javitonino

### LINKED ISSUE #9918: Check deadlocks when creating/deleting auto indices rapidly
What happens:
 1. Long running query on the table
 2. CREATE INDEX concurrently (waits for 1)
 3. DROP INDEX (waits for 3, locks all reads/writes)
 4. New transactions are locked

### PR DESCRIPTION
**Pull request by javitonino**
- Makinf DROP INDEX concurrent fixes #9918. It's slower but does not lock the table.
 - Deleting invalid concurrent indices (a growing problem) fixes #9537. This is done in every run.

### CODE REVIEW

**Comment by javitonino** [comment]
Please CR @juanignaciosl

File: app/services/carto/user_table_index_service.rb [hunk 1]
```diff
@@ -79,8 +84,12 @@ def indexable_column?(column)
       false
     end
 
-    def auto_indices
-      indices.select { |i| i[:name].starts_with?(AUTO_INDEX_PREFIX) }
+    def auto_indices(valid: true)
```
Comment (juanignaciosl) [inline app/services/carto/user_table_index_service.rb]: I just realized that some Spanish sneaked in here :D . Nothing that you should change now, of course.

File: app/services/carto/user_table_index_service.rb [hunk 2]
```diff
@@ -30,10 +30,15 @@ def update_auto_indices
     private
 
     def generate_indices
+      auto_indices(valid: false).each do |idx|
```
Comment (juanignaciosl) [inline app/services/carto/user_table_index_service.rb]: Dropping indexes on creation is strange. Could you remind me how come it's like this, please? When is generation triggered, other than on creation?

File: app/services/carto/user_table_index_service.rb (hunk 1 above)
Comment (javitonino) [inline app/services/carto/user_table_index_service.rb]: [indices is actually the plural of index](http://www.dictionary.com/browse/indices). Although it appears indexes is also valid.

File: app/services/carto/user_table_index_service.rb (hunk 2 above)
Comment (javitonino) [inline app/services/carto/user_table_index_service.rb]: Well, `generate_indices` is the only entry point to the auto indexing routing. It is run in widgets creation/update/deletion, for all affected layers.

In this case, it is scanning for [INVALID indices](https://www.postgresql.org/docs/9.5/static/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY) that may have failed in a previous run in order to delete them so they can be recreated.

File: app/services/carto/user_table_index_service.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/services/carto/user_table_index_service.rb]: 🙊  I should've closed my ignorant mouth! I didn't know and I _wasn't lucky_ with my first search :_D

File: app/services/carto/user_table_index_service.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/services/carto/user_table_index_service.rb]: Ok, I knew what it does, I didn't remember that it runs on every single widget operation 👍

**Pull request merged**